example how to extend the UI with SPI (#521)
keycloak/keycloak-quickstarts: extension/extend-admin-console-spi/README.md

## Extending the admin ui with SPI

This example shows how you can extend the admin ui by implementing the SPI.
You can either add a tab to an existing page or have a complete section in the menu.
By implementing either:

  * `org.keycloak.services.ui.extend.UiPageProvider` for a complete section
  * `org.keycloak.services.ui.extend.UiTabProvider` for a tab

For both options you need to describe the fields you want on the page by using the `org.keycloak.provider.ProviderConfigProperty`
To specify where the tab will appear in the ui, you can add a path valid paths can be found in the `routes.ts` files.

### Running

Run this example by compiling with maven:

```bash
$ mvn install
```

Then copy the generated `extend-admin-ui-1.0-SNAPSHOT.jar` into your <keycloak-server>/provider folder
### Screenshots

Attribute realm tab:
![attribute ream tab](images/img.png)

Todo section
![todo section](images/img_1.png)

Todo detail screen
![todo detail screen](images/img_2.png)

> 

Reviewer: Please update this text to be more descriptive. It need not be too much, at least add explanations how to get to the admin console from the pictures. Feel free to take [some other quickstart](https://github.com/keycloak/keycloak-quickstarts/blob/6231f33555ff80f4d86372a895efe4497c828242/extension/extend-account-console/README.md?plain=1#L35-L71) as an example.

## Extending the admin ui with SPI

This example shows how you can extend the admin ui by implementing the SPI.
You can either add a tab to an existing page or have a complete section in the menu.
By implementing either:

  * `org.keycloak.services.ui.extend.UiPageProvider` for a complete section
  * `org.keycloak.services.ui.extend.UiTabProvider` for a tab

For both options you need to describe the fields you want on the page by using the `org.keycloak.provider.ProviderConfigProperty`

> To specify where the tab will appear in the ui, you can add a path valid paths can be found in the `routes.ts` files.

Reviewer: ```suggestion
To specify where the tab will appear in the ui, you can add a path valid paths can be found in the `routes.ts` files in the admin-ui module of Keycloak project.
```

> This example contains an implementation of both these `AdminUiPage` is a todo section that is added.

Reviewer: ```suggestion
This example contains an implementation of both these. `AdminUiPage` is a todo section that is added.
```

And `ThemeUiTab` is an extra tab in the realm setting that let's you edit the realm attributes.

### Running

Run this example by compiling with maven:

```bash

> $ mvn install

Reviewer: All the other tests have the following command to skip tests (`-DskipTests=true`) and also to take the compilation steps from the main pom (`-Pextension`).
```suggestion
mvn -Pextension clean install -DskipTests=true
```
Is it done here differently on purpose?

```

> Then copy the generated `extend-admin-ui-1.0-SNAPSHOT.jar` into your <keycloak-server>/provider folder

Reviewer: ```suggestion
Then copy the generated `extend-admin-ui-1.0-SNAPSHOT.jar` into your <keycloak-server>/provider folder and run the server using `kc.[sh|bat] start-dev`.
```

### Screenshots

> As you can see from the `getPath` from the `ThemeUiTab` example this page can be found when you navigate to "realm settings"

Reviewer: This assumes knowledge about `getPath` being automatically detected and linked into the _Realm settings_. This may not be necessarily obvious. Is there any reference that would clarify this functionality?

```suggestion
To illustrate how `getPath` method is used, check the `ThemeUiTab` class in this quickstart. The ThemeUi page can now be found when you navigate to _Realm settings_.
```

![attribute ream tab](images/img.png, "Realm attributes tab")

> This is an example of an extra section (see `AdminUiPage`) in the main menu it shows up as the last item in the side menu:

Reviewer: ```suggestion
The `AdminUiPage` class illustrates how to introduce a new section into the main menu in admin console. It adds a Todo section as the last item in the side menu:
```

This example contains an implementation of both these. `AdminUiPage` is a todo section that is added.
And `ThemeUiTab` is an extra tab in the realm setting that let's you edit the realm attributes.

### Running

Run this example by compiling with maven:

```bash
mvn -Pextension clean install -DskipTests=true
```

Then copy the generated `extend-admin-ui-1.0-SNAPSHOT.jar` into your <keycloak-server>/provider folder and run the server using `kc.[sh|bat] start-dev`.

### Screenshots

To illustrate how `getPath` method is used, check the `ThemeUiTab` class in this quickstart. The ThemeUi page can now be found when you navigate to _Realm settings_.
![attribute ream tab](images/img.png, "Realm attributes tab")

The `AdminUiPage` class illustrates how to introduce a new section into the main menu in admin console. It adds a Todo section as the last item in the side menu:
![todo section](images/img_1.png, "Todo section")

After you created an item you can view the details by clicking on it in the list:
![todo detail screen](images/img_2.png, "Todo item details")

### Integration test of the Quickstart

> Make sure you have a Keycloak server running with an admin user in the master realm or use the provided docker image.

Reviewer: This does not fit, the provider needs to be installed into the server first.

> Your Keycloak server should be listening on `http://localhost:8180`. You can archive this by running:

Reviewer: ```suggestion
Your Keycloak server should be listening on `http://localhost:8180`. You can achieve this by running:
```

```
./kc.sh start-dev --http-port=8180
```

> You need to have Chrome browser installed and updated to the latest version. See README for additional details.

Reviewer: ```suggestion
You need to have Chrome browser installed and updated to the latest version. See the root README for additional details.
```

```
mvn clean install -Dextension
```

### Undeploy the quickstart

> Remove sysout listener from your realm in the admin console, delete file providers/event-listener-sysout.jar and restart the server.

Reviewer: Please update the provider name
